In [1]:

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import matplotlib.pyplot as plt

df_train = pd.read_csv('train.csv')

# df_test = pd.read_csv('test.csv')


/home/uday/anaconda2/envs/dato-env/lib/python2.7/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')


In [4]:
place_id = df_train['place_id']
place_id.value_counts()  #8772469670    1849;1623394281    1802;1308450003    1757
# n, bins, patches = plt.hist(place_id, 78, normed=True, facecolor='green', alpha=0.75)
# plt.show()

8772469670    1849
1623394281    1802
1308450003    1757
4823777529    1738
9586338177    1718
9129780742    1716
9544215131    1702
5351837004    1699
4638096372    1699
8610202964    1693
6051554924    1689
7363774279    1689
8607353836    1686
8336299754    1683
5204012807    1680
7230349735    1676
7985287621    1673
4371034975    1673
4993591840    1672
7348940462    1670
7698408658    1669
5971252160    1668
1500208175    1667
8842957841    1661
9919909729    1660
9903336130    1659
6909189260    1658
1628743555    1648
4113278218    1644
3659348746    1631
              ... 
5723178213       1
8770284658       1
9684688654       1
1008054483       1
2150524266       1
5293934773       1
1705380213       1
9663369977       1
3266541302       1
5081250759       1
1474802253       1
6494133525       1
8945287729       1
8806200160       1
3285347166       1
7226358437       1
1232113414       1
5714851113       1
4121531422       1
9200831582       1
4952760562       1
7395434650  

In [5]:
df_train_pop = df_train[df_train['place_id'] == 8772469670]
time_pop = df_train_pop['time']
df_train_pop2 = df_train[df_train['place_id'] == 1623394281]
time_pop2 = df_train_pop2['time']
df_train_pop3 = df_train[df_train['place_id'] == 4113278218]
time_pop3 = df_train_pop3['time']
df_train_pop4 = df_train[df_train['place_id'] == 3659348746]
time_pop4 = df_train_pop4['time']

    

In [15]:
#!/usr/bin/env python
#Example of matplotlib.pyplot.hist
import matplotlib.mlab as mlab

mu, sigma = 100, 15
x = mu + sigma*np.random.randn(10000)

# the histogram of the data
n, bins, patches = plt.hist(x, 50, normed=1, facecolor='green', alpha=0.75)

# add a 'best fit' line
y = mlab.normpdf( bins, mu, sigma)
l = plt.plot(bins, y, 'r--', linewidth=1)

plt.xlabel('Smarts')
plt.ylabel('Probability')
plt.title(r'$\mathrm{Histogram\ of\ IQ:}\ \mu=100,\ \sigma=15$')
plt.axis([40, 160, 0, 0.03])
plt.grid(True)

plt.show()

In [6]:
n, bins, patches = plt.hist(time, normed=True, facecolor='green', alpha=0.75)
plt.show()
# time_pop.hist()

In [7]:
#Note: was getting KeyError: 0L when running code below with time_pop as arg. Had to pass in time_pop.values explicitly 
#to get around it. Something about  API incompatibility with matplotlib and pandas >= 0.13.
#Not sure why the same call worked for time above
# n, bins, patches = plt.hist(time_pop3.values,  facecolor='green', alpha=0.75) #500,000 was the highest
# plt.show()
n, bins, patches = plt.hist([time_pop.values,time_pop2.values,time_pop3.values,time_pop4.values], alpha=0.75) 
plt.show()
#each place_id has a distinct pattern of variability along the time axis

In [8]:
df_train.describe()

row_id             x             y      accuracy          time  \
count  2.911802e+07  2.911802e+07  2.911802e+07  2.911802e+07  2.911802e+07   
mean   1.455901e+07  4.999770e+00  5.001814e+00  8.284912e+01  4.170104e+05   
std    8.405649e+06  2.857601e+00  2.887505e+00  1.147518e+02  2.311761e+05   
min    0.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00  1.000000e+00   
25%    7.279505e+06  2.534700e+00  2.496700e+00  2.700000e+01  2.030570e+05   
50%    1.455901e+07  5.009100e+00  4.988300e+00  6.200000e+01  4.339220e+05   
75%    2.183852e+07  7.461400e+00  7.510300e+00  7.500000e+01  6.204910e+05   
max    2.911802e+07  1.000000e+01  1.000000e+01  1.033000e+03  7.862390e+05   

           place_id  
count  2.911802e+07  
mean   5.493787e+09  
std    2.611088e+09  
min    1.000016e+09  
25%    3.222911e+09  
50%    5.518573e+09  
75%    7.764307e+09  
max    9.999932e+09

In [9]:
#let's next examine if there is geographic locality associated with each place_id
#use manhattan block size 80m x275m to get 125 blocks x 36 blocks in 10KM-squared space
#write function to map (x,y) coordinates to block coordinates
import math
def blockx(x):
    #convert x into a block distance 
    return int(math.floor(x*1000/80))
def blocky(y):
    #convert y into a block distance 
    return int(math.floor(y*1000/275))
    

In [10]:
df_train['blockx'] = map(blockx,df_train['x'])
df_train['blocky'] = map(blocky,df_train['y'])


In [11]:
df_train['blockx'].head()

0      9
1     74
2    103
3     92
4     51
Name: blockx, dtype: int64

In [12]:
blockx(10)

125

In [69]:
blocky(10)

36

In [13]:
df_train00 = df_train[(df_train['blockx'] == 0) & (df_train['blocky'] == 0)]

In [14]:
df_train00.describe()

row_id            x            y     accuracy           time  \
count  4.422000e+03  4422.000000  4422.000000  4422.000000    4422.000000   
mean   1.485416e+07     0.040976     0.155830    87.725690  416688.963591   
std    8.362960e+06     0.021630     0.074700   118.793957  233816.088801   
min    1.490000e+02     0.000000     0.000100     1.000000      45.000000   
25%    7.580357e+06     0.023300     0.104100    26.000000  186273.250000   
50%    1.515110e+07     0.040300     0.150950    63.000000  452430.000000   
75%    2.194103e+07     0.059200     0.219275    81.000000  607890.750000   
max    2.911510e+07     0.079900     0.274900   988.000000  786101.000000   

           place_id  blockx  blocky  
count  4.422000e+03  4422.0  4422.0  
mean   5.866703e+09     0.0     0.0  
std    2.232172e+09     0.0     0.0  
min    1.003836e+09     0.0     0.0  
25%    4.336424e+09     0.0     0.0  
50%    5.445221e+09     0.0     0.0  
75%    8.050783e+09     0.0     0.0  
max    9.932536e+09     0.0     0.0

In [15]:
time_00 = df_train00['time']
n, bins, patches = plt.hist([time_00.values], alpha=0.75) 
plt.show()

In [16]:
place_id00 = df_train00['place_id']
place_id00.value_counts() 

5445221293    766
8050782732    546
8555966805    341
4336423518    271
2477372399    118
4251632872    104
7382144836     98
1006316884     95
8190826815     86
2856338949     86
5357692951     85
7014475222     84
3243336074     72
7065354365     71
8173555001     69
2154052852     62
8815983898     60
8378301865     57
3731899031     56
5349374802     51
7942904509     46
8944067726     43
3271917785     38
5002696932     32
4422663263     30
5938943847     28
5957043142     28
3547886077     25
5765520742     23
8241449659     21
             ... 
3664442439      1
2198399075      1
4771174593      1
6803385643      1
4942361500      1
3214974249      1
1975389431      1
7556775153      1
3077551320      1
1339489481      1
6126650538      1
5998075037      1
7557029004      1
6660617304      1
7864951882      1
6472354795      1
5527864210      1
9447657668      1
1292505932      1
1317237364      1
5387180565      1
3840693543      1
4587422126      1
4982026592      1
5169547607

### the top 30 place_ids account for 3492 rows out of 4422 which is 80% of the data. We therefore prune df_train00 to just these rows

In [18]:
df_train00_top30 = df_train00[df_train00['place_id'].isin([5445221293,
8050782732,
8555966805,
4336423518,
2477372399,
4251632872,
7382144836,
1006316884,
8190826815,
2856338949,
5357692951,
7014475222,
3243336074,
7065354365,
8173555001,
2154052852,
8815983898,
8378301865,
3731899031,
5349374802,
7942904509,
8944067726,
3271917785,
5002696932,
4422663263,
5938943847,
5957043142,
3547886077,
5765520742,
8241449659])]

In [19]:
df_train00_top30.describe()

row_id            x            y     accuracy           time  \
count  3.492000e+03  3492.000000  3492.000000  3492.000000    3492.000000   
mean   1.483209e+07     0.039768     0.158347    86.089633  407572.727377   
std    8.367223e+06     0.021214     0.071923   104.116552  236103.683380   
min    1.490000e+02     0.000000     0.000100     1.000000     410.000000   
25%    7.535368e+06     0.022475     0.106200    31.000000  168889.250000   
50%    1.515110e+07     0.038450     0.153600    64.000000  441495.000000   
75%    2.192826e+07     0.057600     0.218600    83.000000  598082.500000   
max    2.911510e+07     0.079900     0.274700   988.000000  786101.000000   

           place_id  blockx  blocky  
count  3.492000e+03  3492.0  3492.0  
mean   6.020416e+09     0.0     0.0  
std    2.100485e+09     0.0     0.0  
min    1.006317e+09     0.0     0.0  
25%    4.336424e+09     0.0     0.0  
50%    5.445221e+09     0.0     0.0  
75%    8.050783e+09     0.0     0.0  
max    8.944068e+09     0.0     0.0

#### Next we convert time into its constituents: H (hour of day), D (day of week), M (month of year)


In [28]:
def get_hour(time):
    #the same hour of day repeats every 60*24 = 1440 minutes
    return int(math.floor((time%1440)/60))
def get_day(time):
    #the same day of week repeats every 60*24*7 = 10080 minutes
    return int(math.floor((time%10080)/1440))
def get_month(time):
    #the same month of year repeats every 60*24*7*52= 524160 minutes
    return int(math.floor((time%524160)/43680)) #  =524160/12


In [23]:
get_hour(1400)

23

In [27]:
get_day(10080)

0

In [31]:
get_month(524160)

0

In [34]:
df_train00_top30['hour'] = map(get_hour,df_train00_top30['time'])
df_train00_top30['day'] = map(get_day,df_train00_top30['time'])
df_train00_top30['month'] = map(get_month,df_train00_top30['time'])


/home/uday/anaconda2/envs/dato-env/lib/python2.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':
/home/uday/anaconda2/envs/dato-env/lib/python2.7/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app
/home/uday/anaconda2/envs/dato-env/lib/python2.7/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_i

In [33]:
df_train00_top30['hour'].head()

149       8
19019     4
56296    22
63491    14
65582     8
Name: hour, dtype: int64

In [35]:
df_train00_top30['day'].head()

149      5
19019    5
56296    3
63491    2
65582    3
Name: day, dtype: int64

In [36]:
df_train00_top30['month'].head()

149      0
19019    2
56296    3
63491    8
65582    0
Name: month, dtype: int64

#### Use 1-hot encoding to convert hour,day,month to 24+7+12 columns

In [ ]:
def 1_hot_hour(df):
    for i in range(24):
        df['hour'+i] = 1 if 